### PR TITLE
Remove use of DPC++ features deprecated in 2021.4 and open source intel/llvm compiler

### LINKED ISSUE #602: dpctl-capi can no longer be built with Open-source LLVM/SYCL compiler
```
(py39) [12:36:38 ansatnuc04 dpctl]$ clang++ --version
clang version 14.0.0 (https://github.com/intel/llvm 7c17ad9a2bfbf6ff2757f8c7fbd2939841f35fa5)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /localdisk/work/opavlyk/sycl_workspace/llvm/build/bin
```

```
(py39) [12:38:10 ansatnuc04 dpctl]$ git log -1 --oneline
18b56cf (HEAD -> master, tag: 0.10.0rc3, origin/release0.10, origin/master, origin/HEAD) Update CHANGELOG for 0.10 (#590)
```

```bash
python setup.py develop --sycl-compiler-prefix=$(dirname $(dirname `which clang++`))
```

fails with compilation errors due to removal of ``sycl::vector_class``:

```
[ 23%] Building CXX object CMakeFiles/DPCTLSyclInterface.dir/source/dpctl_sycl_device_manager.cpp.o
[ 29%] Building CXX object CMakeFiles/DPCTLSyclInterface.dir/source/dpctl_sycl_device_selector_interface.cpp.o
/localdisk/work/opavlyk/interop/dpctl/dpctl-capi/source/dpctl_sycl_context_interface.cpp:39:36: error: no template named 'vector_class'
DEFINE_SIMPLE_CONVERSION_FUNCTIONS(vector_class<DPCTLSyclDeviceRef>,
                                   ^
/localdisk/work/opavlyk/interop/dpctl/dpctl-capi/source/dpctl_sycl_device_interface.cpp:43:36: error: no template named 'vector_class'
DEFINE_SIMPLE_CONVERSION_FUNCTIONS(vector_class<DPCTLSyclDeviceRef>,
                                   ^
```

I wonder if we could use `vector_class_t` which we would define as `sycl::vector_class` for older compilers, and `std::vector` for newer.

```c++
#if __SYCL_COMPILER_VERSION < 20210925
template <typename T>
using vector_class_t = sycl::vector_class<T>;
#else
template <typename T>
using vector_class_t = std::vector<T>;
#endif
```

and then use `vector_class_t` throughout.

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
Also replaced ``sycl::ONEAPI::filter_selector`` with ``sycl::ext::oneapi::filter_selector``

The packages compiles using open source SYCL compiler after these changes, specifically I used

```
(py39) [13:20:15 ansatnuc04 dpctl]$ clang++ --version
clang version 14.0.0 (https://github.com/intel/llvm 7c17ad9a2bfbf6ff2757f8c7fbd2939841f35fa5)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /localdisk/work/opavlyk/sycl_workspace/llvm/build/bin
```

Only one meaningful failure in the so compiled package:

```
>       dpctl_prog.create_program_from_spirv(q, spirv)

dpctl/tests/test_sycl_program.py:93:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
dpctl/program/_program.pyx:178: in dpctl.program._program.create_program_from_spirv
    cpdef create_program_from_spirv(SyclQueue q, const unsigned char[:] IL,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   raise SyclProgramCompilationError()
E   dpctl.program._program.SyclProgramCompilationError

dpctl/program/_program.pyx:209: SyclProgramCompilationError
--------------------------------------------------------------------------- Captured stderr call ----------------------------------------------------------------------------
The level zero loader dynamic library could not be opened.
ZeModuleCreateFn is invalid.
========================================================================== short test summary info ==========================================================================
FAILED dpctl/tests/test_sycl_program.py::test_create_program_from_spirv_l0 - dpctl.program._program.SyclProgramCompilationError
```

but not sure of the cause. 

This change closes #602, although I think changes here will need adjustments to ensure the package can be built by using both newer and older compiler versions.

### CODE REVIEW

**Comment by oleksandr-pavlyk** [comment]
This change works well with oneAPI 2021.4. I just rebased the branch on top of the master. 

Unless we want to maintain backward compatibility with 2021.3, we could just merge this. 

If maintaining compatibility is an option, we need a type for `vector_class` and for the namespace of `filter_selector` which changed between versions.

**Comment by oleksandr-pavlyk** [comment]
@diptorupd ping

File: dpctl-capi/source/vector_type.hpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+/// \file
+/// This file is a private header of DPCTLSyclInterface to define
+/// vector_class_t, which used to corresponds to sycl::vector_class in earlier
+/// versions of DPC++ and later removed in favor of using std::vector
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#if __SYCL_COMPILER_VERSION < 20210925
```
Comment (oleksandr-pavlyk) [inline dpctl-capi/source/vector_type.hpp:30]: For this preprocessor macro to be available, sycl.hpp needs to be included first.

**Comment by oleksandr-pavlyk** [comment]
Coverage drop is due to change in #606, additional tests should be added in a separate PR.

File: dpctl-capi/source/vector_type.hpp (hunk 1 above)
Comment (oleksandr-pavlyk) [reply dpctl-capi/source/vector_type.hpp:31]: This was fixed, so resolving.

File: dpctl-capi/source/dpctl_sycl_device_selector_interface.cpp [hunk 2]
```diff
@@ -82,8 +82,13 @@ __dpctl_give DPCTLSyclDeviceSelectorRef DPCTLCPUSelector_Create()
 __dpctl_give DPCTLSyclDeviceSelectorRef
 DPCTLFilterSelector_Create(__dpctl_keep const char *filter_str)
 {
+#if __SYCL_COMPILER_VERSION < 20210925
```
Comment (diptorupd) [inline dpctl-capi/source/dpctl_sycl_device_selector_interface.cpp:85]: Do we need this if-def or should just supporting 2021.3 and up be enough?

Comment (oleksandr-pavlyk) [reply dpctl-capi/source/dpctl_sycl_device_selector_interface.cpp:85]: We need it. The change happened between 2021.3 and 2021.4

Comment (diptorupd) [reply dpctl-capi/source/dpctl_sycl_device_selector_interface.cpp:85]: aha right. OK

**Review by diptorupd — APPROVED** [review]
Thanks!

**Pull request merged**